lib/Targets/AIETargetCDODirect.cpp

```diff
@@ -273,48 +273,63 @@ LogicalResult configureBdInBlock(XAie_DevInst &devInst, XAie_DmaDesc &dmaTileBd,
                             qOs, cache, secure);
   }
 
-  // deref here because this is a const iter and the various getters below
-  // aren't const (even though they probably should be...)
   // StringRef FifoMode = disable; // FIXME: when to enable FIFO mode?
-  ShapedType bufferType = bdOp.getBuffer().getType().cast<::mlir::MemRefType>();
-  int bytes = bufferType.getElementTypeBitWidth() / 8;
   int baseAddr = 0;
   if (!targetModel.isShimNOCTile(tileLoc.Col, tileLoc.Row)) {
     auto bufferOp = cast<AIE::BufferOp>(bdOp.getBuffer().getDefiningOp());
-    assert(bufferOp.getAddress().has_value() && "buffer must have address");
+    if (!bufferOp.getAddress())
+      return bufferOp.emitError("buffer must have address assigned");
     baseAddr = bufferOp.getAddress().value();
     if (targetModel.isMemTile(tileLoc.Col, tileLoc.Row))
       baseAddr += BASE_ADDR_A_INCR;
   }
 
   std::optional<llvm::ArrayRef<BDDimLayoutAttr>> dims = bdOp.getDimensions();
-  int lenInBytes = bdOp.getLenValue() * bytes;
-  int basePlusOffset = baseAddr + bdOp.getOffsetValue();
+  int lenInBytes = bdOp.getLenInBytes();
+  int basePlusOffsetInBytes = baseAddr + bdOp.getOffsetInBytes();
+  if (basePlusOffsetInBytes % 4)
+    return bdOp.emitOpError("bd address must be 4B (32b) aligned");
```

Comment:
This seems like it should be in the verifier?

Reply:
Base address is only know after the `aie-assign-buffer-addresses` pass but sure.